src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -188,91 +192,147 @@ private static SequencePosition SeekMultiSegment(ReadOnlySequenceSegment<T> curr
+            Debug.Assert(endSegment is ReadOnlySequenceSegment<T>);
 
-            var current = Unsafe.As<ReadOnlySequenceSegment<T>>(startSegment);
-
-            while (current.Next != null)
+            while (current != null)
             {
-                current = current.Next;
                 if (current == endSegment)
                 {
                     // Found end
                     return;
                 }
+                current = current.Next;
             }
 
             ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private long GetLength(in SequencePosition start, in SequencePosition end)
+        private void CheckEndReachable(ReadOnlySequenceSegment<T> current, object endSegment)
         {
-            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
+            Debug.Assert(endSegment != null);
+            Debug.Assert(endSegment is ReadOnlySequenceSegment<T>);
 
-            object startObject = start.GetObject();
-            object endObject = end.GetObject();
-            if (type == SequenceType.MultiSegment && startObject != endObject)
+            while (current != null)
             {
-                Debug.Assert(startObject != null);
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
-                Debug.Assert(endObject != null);
-                Debug.Assert(endObject is ReadOnlySequenceSegment<T>);
-
-                var startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
-                var endSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(endObject);
-                // (end.RunningIndex + endIndex) - (start.RunningIndex + startIndex) // (End offset) - (start offset)
-                return endSegment.RunningIndex - startSegment.RunningIndex - startIndex + endIndex; // Rearranged to avoid overflow
+                if (current == endSegment)
+                {
+                    // Found end
+                    return;
+                }
+                current = current.Next;
             }
 
-            Debug.Assert(startObject == endObject);
-            // Single segment length
-            return endIndex - startIndex;
+            // Unlike the other CheckEndReachable overload, endSegment is provided by the user, and hence the exception is different
+            ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void BoundsCheck(in SequencePosition start, in SequencePosition end)
+        private void BoundsCheck(in SequencePosition start)
         {
-            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
+            uint sliceStartIndex = (uint)GetIndex(start);
```

Comment:
Why do you need casts to uint/ulong?

Reply:
It is necessary to correctly deal with underflow/overflow which lets us optimize the InRange method to be a single comparison.

Reply:
> It is necessary to correctly deal with underflow/overflow which lets us optimize the InRange method to be a single comparison.

I think you can use `int` and `long` as InRange params and  `uint` and `ulong`  inside InRange.